zulipterminal/ui_tools/boxes.py

```diff
@@ -476,14 +501,24 @@ def soup2markup(self, soup: Any, **state: Any) -> List[Any]:
                     # Prepend org url to convert it to an absolute link
                     link = urljoin(self.model.server_url, link)
 
-                if link == text:
-                    # If the link and text are same
-                    # usually the case when user just pastes
-                    # a link then just display the link
-                    markup.append(('msg_link', text))
+                text = text if text else link
+
+                # Detect duplicate links to save screen real estate.
+                if link not in self.message_links:
+                    self.message_links[link] = (text, self.link_index)
+                    self.link_index += 1
                 else:
-                    markup.append(
-                        ('msg_link', '[' + text + ']' + '(' + link + ')'))
+                    saved_text, saved_link_index = self.message_links[link]
+                    if saved_text != text:
+                        self.message_links[link] = (
+                            '{}, {}'.format(saved_text, text), saved_link_index
+                        )
```

Comment:
You don't have a test for this?

Reply:
The 'saved text' is not used anywhere currently and is for the popup (where we'll show `1: Text \n Link`). However, the code is still useful to have so that we don't miss the 'saved text' even though we're not using it.

I have added extra test cases that are remotely related.